src/ws/session-utils.c

```diff
@@ -873,22 +873,40 @@ authorize_logger (const char *data)
 FILE *
 open_memfd (const char *name)
 {
-  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING);
+  int fd = memfd_create ("cockpit login messages", MFD_ALLOW_SEALING | MFD_CLOEXEC);
 
   if (fd == -1)
     return NULL;
 
   return fdopen (fd, "w");
 }
 
-bool
-seal_memfd (FILE *memfd)
+int
+seal_memfd (FILE **memfd)
 {
-  if (fflush (memfd) != 0)
-    return false;
+  char fd_name[] = "/proc/self/fd/xxxxxx";
+  int fd;
+
+  if (fflush (*memfd) != 0)
+    return -1;
+
+  fd = fileno (*memfd);
 
   const int seals = F_SEAL_SHRINK | F_SEAL_GROW | F_SEAL_WRITE;
-  return fcntl (fileno (memfd), F_ADD_SEALS, seals) == 0;
+  if (fcntl (fd, F_ADD_SEALS, seals) != 0)
+    return -1;
+
+  int r = snprintf (fd_name, sizeof fd_name, "/proc/self/fd/%d", fd);
```

Comment:
This looks rather hackish.. Could `fcntl(fd, F_SETFL, O_RDONLY)` possibly work here? Or, as `memfd` is already a stream, you could use `fdopen(fd, "r")` and return its fileno()?

Reply:
`fdopen()` is a pure libc stdio wrapper: it doesn't reopen the underlying fd.

meanwhile `fcntl(F_SETFL)` is explicitly documented:

```
       F_SETFL (int)
              Set the file status flags to the value specified by
              arg.  File access mode (O_RDONLY, O_WRONLY, O_RDWR)
              and  file  creation  flags  (i.e., O_CREAT, O_EXCL,
              O_NOCTTY, O_TRUNC) in arg are ignored.
```

Reply:
OK, fair enough. Thanks! What a gap in Linux..